### ISSUE TITLE
Registry freezes regularly whilst fetching image from S3

### ISSUE THREAD

**Issue by pwaller** [docker/docker-registry#866]
Regularly docker pull fails by either hanging or crashing out.

I've narrowed the problem down, I can reproduce it reliably with a curl to `http://.../v1/images/.../layer` (which is ~100MiB) with no proxies in the way. The frequency seems to vary. Sometimes it is 1 in 15, sometimes it is 1 in 2.

The problem has a similar feel to #815.

```
Traceback (most recent call last):
File "/usr/local/lib/python2.7/dist-packages/gevent/greenlet.py", line 327, in run
  result = self._run(*self.args, **self.kwargs)
File "/usr/local/lib/python2.7/dist-packages/docker_registry/core/boto.py", line 75, in _fetch_part
  boto_key.get_contents_to_file(f, headers={'Range': brange})
File "/usr/local/lib/pyth  on2.7/dist-packages/boto/s3/key.py", line 1648, in get_contents_to_file
  response_headers=response_headers)  
File "/usr/local/lib/python2.7/dist-packages/boto/s3/key.py", line 1480, in get_file
  query_args=None)  
File "/usr/local/lib/python2.7/dist-packages/boto/s3/key.py", line 1533, in _get_file_internal
  for bytes in self:  
File "/usr/local/lib/python2.7/dist-packages/boto/s3/key.py", line 386, in next
  data = self.resp.read(self.BufferSize)
File "/usr/local/lib/python2.7/dist-packages/boto/connection.py", line 413, in read
  return http_client.HTTPResponse.read(self, amt)
File "/usr/lib/python2.7/httplib.py", line 567, in read
  s = self.fp.read(amt)
File "/usr/lib/python2.7/socket.py", line 380, in read
  data = self._sock.recv(left)
File "/usr/local/lib/python2.7/dist-packages/gevent/ssl.py", line 208, in recv
  return self.read(buflen)
File "/usr/local/lib/python2.7/dist-packages/gevent/ssl.py", line 110, in read
  return self._sslobj.read(len)
error: [Errno 104] Connection reset by peer
<Greenlet at 0x7f9a8a95f410: <bound method ParallelKey._fetch_part of <docker_registry.core.boto.ParallelKey object at 0x7f9a8a8a8250>>('/tmp')
```

Docker client output looks like this:
```
Error pulling image (latest) from ***/***, endpoint: https://***/v1/, ApplyLayer exit status 1 flate: corrupt input before offset 13494754
Error pulling image (latest) from ***/***, ApplyLayer exit status 1 flate: corrupt input before offset 13494754
Container run failed: Error pulling image (latest) from ***/***, ApplyLayer exit status 1 flate: corrupt input before offset 13494754
```

**Comment by dmp42** [comment]
@pwaller any news on trying without AWS_ENCRYPT and AWS_SECURE?

**Comment by pwaller** [comment]
We haven't done much over the holiday period. Glad to hear IAM authentication is in the rewrite.

Please re-ping if you come across this issue in greater than 10 days from now, since I should continue to experience it and be frustrated enough to do the AWS_ENCRYPT experiment if it is still a problem.

**Comment by pwaller** [comment]
Tidying up my personal issues list, so closing this. Please create a new issue if you're still interested in tracking it.

**Comment by dmp42** [comment]
I believe this was fixed by #961
